src/basic/string-util.rs

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+use std::ffi::{CStr, CString};
+use std::os::raw::c_char;
+use std::ptr;
+
+/// Extract the i'nth line from the specified string. Returns > 0 if there are more lines after that,
+/// and == 0 if we are looking at the last line or already beyond the last line. As special
+/// optimization, if the first line is requested and the string only consists of one line we return
+/// NULL, indicating the input string should be used as is, and avoid a memory allocation for a very
+/// common case.
+#[no_mangle]
+pub extern "C" fn string_extract_line(s: *const c_char, i: usize, ret: *mut *mut c_char) -> i32 {
+    if ret.is_null() {
+        panic!("string_extract_line(): 'ret' cannot be NULL");
+    }
```

Comment:
i ifgure the 4ch indenting is rust-native styling?

Reply:
hmm, not too much clue about rust here, but there must be a way to encode this in the function call signature, no?

Reply:
> i ifgure the 4ch indenting is rust-native styling?

https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/guide.md#indentation-and-line-width

> hmm, not too much clue about rust here, but there must be a way to encode this in the function call signature, no?

Quite possibly, but I don't know. First time ever I wrote Rust :-)

Reply:
Rust person here; can I offer some help? 😀

> i ifgure the 4ch indenting is rust-native styling?

Yes

> hmm, not too much clue about rust here, but there must be a way to encode this in the function call signature, no?

Not really. An actual Rust-y version of this function would look somewhat like this:

```rust
fn string_extract_line(s: &str, line_index: usize) -> Option<&str> {
    s.lines().nth(line_index)
}
```
(Note that this always returns a slice of the original string and never does a copy.)

But this function is instead written to use C types (and C ABI) and be callable from external C. So it's rather pointless to try to enforce anything in its Rust-side signature, as C is still going to see its own definition. That, and also raw pointers (as opposed to references) can be null, dangling and unaligned.

On the other hand, since this function is already unsafe to call (it dereferences raw pointers! — and by the way, it should be marked `unsafe` for this reason), and since the (old) C version of this function check doesn't appear to check the pointer for being null, it's OK to just say this is similarly a precondition here.

Reply:
> Rust person here; can I offer some help? 

Yes, please! Pushed a new version, with some improvements

Reply:
Yes, this is much closer to what I would have written.

I would write it like this:

```rust
pub unsafe extern "C" fn string_extract_line(s: *const c_char, i: usize, ret: *mut *mut c_char) -> c_int {
    let s = CStr::from_ptr(s).to_bytes();
    let line = s.split(|&b| b == b'\n').nth(i);
    match line {
        Some(line) if line == s => {
            debug_assert_eq!(i, 0);
            // Special case: return null, to let the caller
            // know to reuse the input string.
            *ret = ptr::null_mut();
            0
        }
        Some(line) => {
            // Make a copy of the substring.
            let owned_line = CString::new(line)
                .expect("Line can't have a zero byte inside, since it's a substring of a preexisting C string");
            *ret = owned_line.into_raw();
            let have_more_lines = line.as_ptr_range().end.add(1) < s.as_ptr_range().end;
            if have_more_lines {
                1
            } else {
                0
            }
        }
        None => {
            // No line with this index.
            *ret = CString::new("").unwrap().into_raw();
            0
        }
    }
}
```
([playground with some tests](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ca729c1478cd21893ae6f8f5ebde4ac2))

i.e. I'd skip converting to a Rust `str`/`String` at all, and stay within `CStr`/`CString`/`[u8]`; but I'm not saying my version is necessarily better.

---

One caveat to be aware of: [`CString::into_raw()` docs](https://doc.rust-lang.org/stable/std/ffi/struct.CString.html#method.into_raw) warn against freeing the resulting pointer with libc `free()`. The reason for this is that it's possible to configure Rust standard library to use a different allocator, so various Rust libraries should not implicitly rely on the libc allocator being used. But in your case I guess you control how the entire thing is built and know what you're doing.

Reply:
> i ifgure the 4ch indenting is rust-native styling?

>> Yes

@bugaevc I took a look at a couple of Rust projects (to figure out how they are tested in general) and my understanding is that rules like that can be automatically checked and enforced by running `rustfmt` and `rustclippy`. I wonder if this should also be somehow done by `meson` automagically or if there is another way to run those tools as part of the systemd CI?

Reply:
Yes; rustfmt is an automatic formatter tool (like clang-format), and clippy is the (aptly named) linter.

To the best of my knowledge, Meson doesn't automatically run either (nor do the names appear in the Meson codebase). It should be possible to add them to Meson as a custom target though.

Do you automatically run clang-format in CI? I've been formatting things manually (well, with a little help from my editor) in my PRs.

Reply:
I'd also add that it appears it doesn't seem to be possible to go far in terms of testing without `cargo` so I'd probably reconsider the "no cargo" policy

Reply:
> Do you automatically run clang-format in CI?

As far as I can remember there was an attempt to do that but at the time clang-format couldn't cover all the systemd use cases. I'm not sure if anything has changed since then though. Though I think it's a bit hard to introduce auto-formatters to code bases that have been around for a while :-) With Rust it should be easier I think because there isn't any legacy Rust stuff.

Reply:
> I'd also add that it appears it doesn't seem to be possible to go far in terms of testing without `cargo`

What do you mean?

The `cargo test` command is mostly a wrapper around passing the `--test` flag to rustc and then running the resulting test executable. It also causes Cargo to enable "dev-dependencies" (those needed for tests and the like, but not for the actual code) and do some other things, but these are Cargo specifics; if you don't use Cargo you can ignore them.

Moreover, you could even forego the built-in testing framework (the one `#[test]` refers to) and just do it the usual Meson way, i.e. by writing separate executables that test whatever functionality.

Reply:
> What do you mean?

What I was trying to say is that everything is done via `cargo` (at least in projects I've taken a look at) and it isn't clear to me how I can replace for example `cargo fuzz` with `rustc`.

Reply:
It is true that virtually everything in the Rust ecosystem uses Cargo because it's terribly convenient; but Rust itself is not tied to Cargo, and Cargo is "just" the package manager and the build system. It's always possible to do things without Cargo, if you want to.

If you want to depend on other crates from crates.io, then perhaps it would be hard to do without Cargo. If you just want to compile some Rust and run some tests, Cargo is in no way mandatory. Though you will run into a lot of uncharted territory, because again most projects do use Cargo and so all the tutorials, etc. assume you do too.

Reply:
Just to clarify by "testing" I didn't mean just running a bunch executables with `cargo test`. I'm talking about various analyzers (like Coverity and LGTM for C), fuzzers and so on. I think Rust code should be tested thoroughly as well.

Reply:
> I think Rust code should be tested thoroughly as well.

Sure.

> I'm talking about various analyzers (like Coverity and LGTM for C), fuzzers and so on.

Well, I again admit I don't have any experience with using those tools with Rust, but Rust should not be much different from C in the regard. You pass the appropriate flags to the compiler (and perhaps Meson handles this for you similarly to how Cargo does), and then you invoke whatever analyzer tool. As discovered below, you'd likely need libstd built with the same flags, but that should be it.

But this is not my area of expertise (I do use Cargo for all my Rust projects...); better ask Meson developers for what they recommend.

Reply:
So there are 2 ways that cargo runs commands

 1. The subcommand is built into cargo. In this case it is intended that anything you can do with cargo you can also do by calling rustc or other binaries like rustdoc directly.
 2. Cargo also provides an extension mechanism so arbitrary commands can masquerade as cargo subcommands. Basically, if some executable with a name like `cargo-mytool` or `cargo-mytool.exe` on windows exists, then running `cargo mytool` will run your tool and pass command line parameters in.

It should be possible to run those programs simply by calling them directly (e.g. `cargo-fuzz --my --args`).

Reply:
A cursory search of [`cargo-fuzz`](https://github.com/rust-fuzz/cargo-fuzz/) suggests that there's no requirement for `cargo` when running the tool (although building it does require cargo).

Reply:
In the second case the command very likely uses cargo internally. Depending on the tool, the cargo subcommand either has a hard dependency on cargo or just uses cargo to build everything with the necessary flags and for example overrides the rustc used by cargo or gets the necessary information about the target crate after building all dependencies and then invokes the main tool. `cargo clippy` for example uses `RUSTC_WORKSPACE_WRAPPER` to run `clippy-driver` instead of rustc for all crates inside the current workspace.

Reply:
I think I more or less figured out how to extract the part of "cargo fuzz" dealing with the compiler options like `-Cpasses=sancov`, `-Cllvm-args=-sanitizer-coverage-level=4` and so on I need. It kind of even works in the sense that I was able to run the systemd fuzz targets built with both clang and rustc. So I agree that it's possible to kind of bring cargo plugins to the systemd build system. The problem is that borrowing code like that isn't maintainable. Someone would have to keep track of what changed where and update `systemd` accordingly.